app/services/umrdr/date_coverage_service.rb

```diff
@@ -0,0 +1,130 @@
+        reverse_interval!(interv) if interval_reversed?(interv) 
+        make_begin_params(interv.begin).merge(make_end_params(interv.end))
+      end
+
+      private
+
+      # Determines if an intervals endings are reverse.  Specifically, where the begin 
+      # date is later that the end date.  An interval with reversed starting and ending
+      # dates will have from and to attributes, but begin and end will be nil
+      # == Parameters:
+      # interv::
+      #   [EDTF::Interval] The interval to be tested. 
+      # == Returns:
+      #   [Boolean] True if the interval boundaries were reversed
+      def interval_reversed?(interv)
+        return false if interv.from == EDTF_OPEN_BEGIN || interv.to == EDTF_OPEN_END 
+        (interv.from && !interv.begin) && (interv.to && !interv.end)
+      end
+
+      # Reverse the begin and end of an interval.  This will bork a well formed interval.
+      # == Parameters:
+      # interv::
+      #   [EDTF::Interval] The interval to be reversed. 
+      # == Returns:
+      #   [EDTF::Interval] Interval with from and to in the reverse order from the input interval.
+      def reverse_interval!(interv)
+        tmp = interv.to
+        interv.to = interv.from
+        interv.from = tmp
+      end
+
+      def make_begin_params(date)
+        hsh = date_to_params date
+        {date_coverage_begin_year:  hsh[:year],
+         date_coverage_begin_month: hsh[:month],
+         date_coverage_begin_day:   hsh[:day]}
+      end
+
+      def make_end_params( date )
+        hsh = date_to_params date
+        {date_coverage_end_year:  hsh[:year],
+         date_coverage_end_month: hsh[:month],
+         date_coverage_end_day:   hsh[:day]}
+      end
+
+      # Given edtf date
+      # return hash with possible keys :year, :month, :day
+      def date_to_params( date )
+        return {} unless date
+        hsh = case date.precision
+        when :year
+          {year: date.year}
+        when :month
+          {year: date.year, month: date.month}
+        when :day
+          {year: date.year, month: date.month, day: date.day}
+        end
+        hsh[:year] = year_to_s hsh[:year]
+        hsh.each{|k,v| hsh[k] = v.to_s}
+      end
+
+      def year_to_s(val)
+        if val.negative?
+          sprintf('-%04d', val.abs)
+        else
+          sprintf('%04d', val)
+        end
+      end
+
+      # Given input
+      # Return the corresponding integer or nil
+      def integerize(input)
+        Integer(input,10) rescue nil
+      end
+
+      # Until 8601-2 gets finalized, edtf ranges do not accept :open for begin.
+      def make_begin_date(y = nil, m = nil, d = nil)
+        make_date(y,m,d) || EDTF_OPEN_BEGIN
+      end
+
+      def make_end_date(y = nil, m = nil, d = nil)
+        make_date(y,m,d) || EDTF_OPEN_END
+      end
+
+      # Takes year, month, and day integers.
+      # Returns edtf date or nil
+      def make_date(y = nil, m = nil, d = nil)
+        dt = case
```

Comment:
It makes sense, it just looks a little strange. I'd probably do `elsif` instead and change the variable names to `year`, `month` and `day`. If you were feeling particularly sassy you could write something like:

```ruby
def make_date(*values)
  Date.new(*values).tap do |date|
    case values.size
    when 3
      date.day_precision!
    when 2
      date.month_precision!
    when 1
      date.year_precision!
    end
  end
end
```

... and if you are feeling **really** sassy, that is probably something the Date initializer should handle OR something that a decorator around the Date class should handle.